### PR TITLE
Restore appimage_get_elf_size implementation on libappimage_share

### PR DESCRIPTION
**Pull request by azubieta**
This is required function is required by the AppImageKit runtime target, therefore, we must keep a bare C implementation.

### CODE REVIEW

**Comment by azubieta** [comment]
That method is written in C++, runtime uses only C.

**Comment by TheAssassin** [comment]
Ah, makes sense. We should consider moving the method out of libappimage into the runtime code base, though. I don't want really redundant code in here.

**Comment by azubieta** [comment]
I guess that once we finish the whole "libappimage" rewrite thing we will have to take a look at the AppImageKit repo in order to do improvements like this one.

File: src/libappimage_shared/elf.c [hunk 1]
```diff
@@ -117,6 +117,41 @@ static off_t read_elf64(FILE* fd)
...
+			ehdr.e_ident[EI_DATA]);
+		return -1;
+	}
+	if (ehdr.e_ident[EI_CLASS] == ELFCLASS32) {
+		size = read_elf32(fd);
+	} else if (ehdr.e_ident[EI_CLASS] == ELFCLASS64) {
+		size = read_elf64(fd);
+	} else {
+		fprintf(stderr, "Unknown ELF class %u\n", ehdr.e_ident[EI_CLASS]);
+		return -1;
```
Comment (TheAssassin) [inline src/libappimage_shared/elf.c]: `fd` is leaked here. Should be fixed before merging, even if this is legacy code that will be removed soon-ish.

File: src/libappimage_shared/elf.c [hunk 2]
```diff
@@ -117,6 +117,41 @@ static off_t read_elf64(FILE* fd)
 	return sht_end > last_section_end ? sht_end : last_section_end;
 }
 
+ssize_t appimage_get_elf_size(const char* fname) {
+    off_t ret;
+    FILE* fd = NULL;
```
Comment (TheAssassin) [inline src/libappimage_shared/elf.c]: No need to set this to `NULL`. Should be fixed.

File: src/libappimage_shared/elf.c [hunk 3]
```diff
@@ -117,6 +117,41 @@ static off_t read_elf64(FILE* fd)
...
+    off_t size = -1;
+
+	fd = fopen(fname, "rb");
```
Comment (TheAssassin) [inline src/libappimage_shared/elf.c]: Broken indentation.

File: src/libappimage_shared/elf.c [hunk 4]
```diff
@@ -117,6 +117,41 @@ static off_t read_elf64(FILE* fd)
...
+	if (fd == NULL) {
+		fprintf(stderr, "Cannot open %s: %s\n",
+			fname, strerror(errno));
+		return -1;
+	}
+	ret = fread(ehdr.e_ident, 1, EI_NIDENT, fd);
+	if (ret != EI_NIDENT) {
+		fprintf(stderr, "Read of e_ident from %s failed: %s\n",
+			fname, strerror(errno));
+		return -1;
```
Comment (TheAssassin) [inline src/libappimage_shared/elf.c]: `fd` is leaked here.

File: src/libappimage_shared/elf.c [hunk 5]
```diff
@@ -117,6 +117,41 @@ static off_t read_elf64(FILE* fd)
...
+	}
+	if ((ehdr.e_ident[EI_DATA] != ELFDATA2LSB) &&
+		(ehdr.e_ident[EI_DATA] != ELFDATA2MSB)) {
+		fprintf(stderr, "Unknown ELF data order %u\n",
+			ehdr.e_ident[EI_DATA]);
+		return -1;
```
Comment (TheAssassin) [inline src/libappimage_shared/elf.c]: `fd` is leaked here.

**Comment by TheAssassin** [comment]
No need for this, actually, thinking about it. AppImageKit shall continue to use the 0.1.x branch until we start to really work on it. This is just not necessary IMO. Closing.

**Comment by azubieta** [comment]
If so we should remove (or at least not install) from the 0.2 all the binaries that are not libappimage_0.2. Otherwise, we will be breaking their API.

**Comment by TheAssassin** [comment]
Your statement doesn't make much sense. Please elaborate.

**Comment by azubieta** [comment]
From the libappimage project are created several binaries. But only the `libappimage` is versioned. This change solves an API breaking in the `libappimage_shared` binary. My point is that if we are not going to provide backward compatibility to the other binaries we must not install them to avoid breaking the software of potential users.

**Comment by TheAssassin** [comment]
Well, that's a bit of an exaggerated reaction. All that prevents us from using this in AppImageKit is that one function. This function doesn't have to be here, though. It can be moved to AppImageKit. Then, it will just work normally.

**Comment by azubieta** [comment]
I guess that I would be good to review all those binaries and see if they should remain as part of the libappimage project or moved elsewere. Opening an issue for it.

**Pull request closed without merging**